src/include/duckdb/function/table/arrow/arrow_type_info.hpp

```diff
@@ -0,0 +1,128 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/function/table/arrow/arrow_type_info.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/common/types.hpp"
+#include "duckdb/common/unordered_map.hpp"
+#include "duckdb/common/vector.hpp"
+#include "duckdb/common/unique_ptr.hpp"
+#include "duckdb/function/table/arrow/enum/arrow_type_info_type.hpp"
+#include "duckdb/function/table/arrow/enum/arrow_datetime_type.hpp"
+#include "duckdb/function/table/arrow/enum/arrow_variable_size_type.hpp"
+#include "duckdb/common/enum_util.hpp"
+
+namespace duckdb {
+
+class ArrowType;
+
+struct ArrowTypeInfo {
+public:
+	explicit ArrowTypeInfo(ArrowTypeInfoType type);
+	virtual ~ArrowTypeInfo();
+
+	ArrowTypeInfoType type;
+
+public:
+	template <class TARGET>
+	TARGET &Cast() {
+		D_ASSERT(dynamic_cast<TARGET *>(this));
+		if (type != TARGET::TYPE) {
+			throw InternalException("Failed to cast ArrowTypeInfo, type mismatch (expected: %s, got: %s)",
+			                        EnumUtil::ToString(TARGET::TYPE), EnumUtil::ToString(type));
+		}
+		return reinterpret_cast<TARGET &>(*this);
+	}
+	template <class TARGET>
+	const TARGET &Cast() const {
+		D_ASSERT(dynamic_cast<const TARGET *>(this));
+		if (type != TARGET::TYPE) {
+			throw InternalException("Failed to cast ArrowTypeInfo, type mismatch (expected: %s, got: %s)",
+			                        EnumUtil::ToString(TARGET::TYPE), EnumUtil::ToString(type));
+		}
+		return reinterpret_cast<const TARGET &>(*this);
+	}
+};
+
+struct ArrowStructInfo : public ArrowTypeInfo {
+public:
+	static constexpr const ArrowTypeInfoType TYPE = ArrowTypeInfoType::STRUCT;
+
+public:
+	explicit ArrowStructInfo(vector<unique_ptr<ArrowType>> children);
+	~ArrowStructInfo() override;
+
+public:
+	idx_t ChildCount() const;
+	const ArrowType &GetChild(idx_t index) const;
+	const vector<unique_ptr<ArrowType>> &GetChildren() const;
+
+private:
+	vector<unique_ptr<ArrowType>> children;
+};
+
+struct ArrowDateTimeInfo : public ArrowTypeInfo {
+public:
+	static constexpr const ArrowTypeInfoType TYPE = ArrowTypeInfoType::DATE_TIME;
+
+public:
+	explicit ArrowDateTimeInfo(ArrowDateTimeType size);
+	~ArrowDateTimeInfo() override;
+
+public:
+	ArrowDateTimeType GetDateTimeType() const;
+
+private:
+	ArrowDateTimeType size_type;
+};
+
+struct ArrowStringInfo : public ArrowTypeInfo {
+public:
+	static constexpr const ArrowTypeInfoType TYPE = ArrowTypeInfoType::STRING;
+
+public:
+	explicit ArrowStringInfo(ArrowVariableSizeType size);
+	explicit ArrowStringInfo(idx_t fixed_size);
+	~ArrowStringInfo() override;
+
+public:
+	ArrowVariableSizeType GetSizeType() const;
+	idx_t FixedSize() const;
+
+private:
+	ArrowVariableSizeType size_type;
+	idx_t fixed_size;
+};
+
+struct ArrowListInfo : public ArrowTypeInfo {
```

Comment:
I see you're going for a partitioning based on logical types here, which I think is very nice. But I would advise against putting fixed-size lists on the same bundle — you can convert any fixed-size list to a variable-length list ~array~vector, but not the other way around. List and list-views conversions work in both ways.

Besides that, the implementation of operations on fixed-size lists can be totally different (much simpler!) than operations on lists and list-views. One thing that list view unlocks that lists don't have is out-of-order appending of values, but the way you achieve that for list-views is so different from the way you do that on fixed-size lists, that sharing logic doesn't make much sense.

Reply:
That said, I think you should probably ignore what I said above because DuckDB seems to unify everythin to variable-length list vectors.

Reply:
This ArrowTypeInfo is just used to contain information extracted from analyzing the ArrowSchema in the bind step